### ISSUE TITLE
Improve memory management

### ISSUE THREAD

**Issue by sbrunk** [sbrunk/storch#5]
We can't rely on the garbage collector to free native memory in time as it is too slow and unpredictable. This is especially true for GPU tensors because usually we want to utilize the constrained GPU memory as much as possible without allocating too much and crashing the process.

So we need a solution using reference counting and explicit deallocation, not only for single tensors, but for groups as well.
[PointerScope](http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/PointerScope.html) ([example](http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/)) from JavaCPP is a working way to achieve that:

```scala
trainDL.foreach { (input, label) =>
  optimizer.zeroGrad()
  Using.resource(new PointerScope()) { p =>
    val pred = model(input.to(device))
    val loss = lossFn(pred, label.to(device))
    loss.backward()
    optimizer.step()
  }
```

This will deallocate all tensors created inside the `Using` block as soon as it is out of scope. One question is if we want to expose `PointerScope`.

If we want to keep i.e. the loss tensor for stats, we need to manually increase the reference counter of the native loss tensor. I'm not sure how well that would work and look like in practice, needs investigation.

scala-torch seems to be doing something similar with their ReferenceManager, but using implicits/context params which should be worth a look as well. See https://github.com/microsoft/scala_torch#memory-management

**Comment by jypma** [comment]
I've been having luck with the following small abstraction over `PointerScope`:
```scala
trait HeapExternal[T] {
  def toPointer(t: T): Pointer
}

object HeapExternal {
  given HeapExternal[Unit] with { def toPointer(u: Unit) = null }
  given HeapExternal[Pointer] with { def toPointer (p: Pointer) = p }
  given tensor[S <: Tuple, T <: DType, D <: Device]: HeapExternal[Tensor[S, T, D]] with { def toPointer (t: Tensor[S, T, D]) = t.native }

  /** Runs the given block, removing all allocated pointers afterwards. T itself however is preserved, and put into any
    * enclosing scope (if one is open).
    */
  def scoped[T: HeapExternal](block: =>T): T = {
    val r = Using.resource(new PointerScope()) { scope =>
      val result = block
      val ptr = summon[HeapExternal[T]].toPointer(result)
      if (ptr != null) {
        ptr.retainReference()
      }
      result
    }
    val ptr = summon[HeapExternal[T]].toPointer(r)
    val scope = PointerScope.getInnerScope()
    if (scope != null && ptr != null) {
      scope.attach(ptr)
    }
    if (ptr != null) {
      ptr.releaseReference()
    }
    r
  }
}
```

This allows the user to write code like:

```scala
scoped {
  for (i <- 0.until(trainingCount)) {
    val resultTensor = scoped { doSomeTraining(model) } // memory from in here is released, but not the result.
  }
} // all results are released here as well.
```

I think I like it better than requiring implicits to each and every method argument. Especially since the internal `Pointer` constructor already refers to the `ThreadLocal` anyway; we'd not be saving performance there either.
